cluster-autoscaler/utils/taints/taints.go

```diff
@@ -355,33 +360,59 @@ func CleanTaints(node *apiv1.Node, client kube_client.Interface, taintKeys []str
 	}
 }
 
+// deletionCandidateStalenessCondition returns a function that checks if a node's deletion candidate time has reached the specified TTL.
+func deletionCandidateStalenessCondition(deletionCandidateTTL time.Duration) func(*apiv1.Node) bool {
+	return func(node *apiv1.Node) bool {
+		if deletionCandidateTTL == 0 {
+			return true
+		}
+		markedForDeletionTime, err := GetDeletionCandidateTime(node)
+		if err != nil {
+			klog.Warningf("Error while getting DeletionCandidate time for node %v: %v", node.Name, err)
+			return true
+		}
+		if markedForDeletionTime == nil {
+			return true
+		}
+		if time.Since(*markedForDeletionTime) > deletionCandidateTTL {
+			timeSinceDeletion := time.Since(*markedForDeletionTime)
+			klog.V(4).Infof("Node %v has stale %v taint: the time is %v (%v ago)", node.Name, DeletionCandidateTaint, *markedForDeletionTime, timeSinceDeletion)
+			return false
+		}
+		return true
+	}
+}
+
 // CleanAllToBeDeleted cleans ToBeDeleted taints from given nodes.
 func CleanAllToBeDeleted(nodes []*apiv1.Node, client kube_client.Interface, recorder kube_record.EventRecorder, cordonNode bool) {
-	CleanAllTaints(nodes, client, recorder, []string{ToBeDeletedTaint}, cordonNode)
+	CleanAllTaints(nodes, client, recorder, ToBeDeletedTaint, cordonNode)
 }
 
-// CleanAllDeletionCandidates cleans DeletionCandidate taints from given nodes.
-func CleanAllDeletionCandidates(nodes []*apiv1.Node, client kube_client.Interface, recorder kube_record.EventRecorder) {
-	CleanAllTaints(nodes, client, recorder, []string{DeletionCandidateTaint}, false)
+// CleanStaleDeletionCandidates cleans DeletionCandidate taints from given nodes.
+func CleanStaleDeletionCandidates(nodes []*apiv1.Node, client kube_client.Interface, recorder kube_record.EventRecorder, deletionCandidateTTL time.Duration) {
+	CleanAllTaints(nodes, client, recorder, DeletionCandidateTaint, false, deletionCandidateStalenessCondition(deletionCandidateTTL))
 }
 
 // CleanAllTaints cleans all specified taints from given nodes.
-func CleanAllTaints(nodes []*apiv1.Node, client kube_client.Interface, recorder kube_record.EventRecorder, taintKeys []string, cordonNode bool) {
+func CleanAllTaints(nodes []*apiv1.Node, client kube_client.Interface, recorder kube_record.EventRecorder, taintKey string, cordonNode bool, conditions ...func(*apiv1.Node) bool) {
 	for _, node := range nodes {
 		taintsPresent := false
-		for _, taintKey := range taintKeys {
-			taintsPresent = taintsPresent || HasTaint(node, taintKey)
-		}
+		taintsPresent = taintsPresent || HasTaint(node, taintKey)
```

Comment:
At this point you don't need the initial assignment of `false` anymore. This can be simplified to:

```
if !HasTaint(node, taintKey) {
  continue
}
```